Add support for CAP_PROP_ORIENTATION_AUTO to AVFoundation backend

tailsu opened the pull request:
This is an extension to #15499. I pulled some of the code up, so now it would be somewhat easier to add support for orientation metadata into the other backends as well.

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

asmorkalov commented:
Thanks a lot for the contribution. It makes sense to extend existing FFmpeg targeted test for Apple platform too. See https://github.com/opencv/opencv/blob/4.x/modules/videoio/test/test_ffmpeg.cpp#L479.

alalek commented on modules/videoio/src/cap_interface.hpp:
```diff
@@ -249,21 +274,58 @@ class LegacyCapture : public IVideoCapture
     double getProperty(int propId) const CV_OVERRIDE
     {
-        return cap ? cap->getProperty(propId) : 0;
+        switch(propId)
+        {
+            case cv::CAP_PROP_ORIENTATION_AUTO:
+                return static_cast<double>(autorotate);
+
+            default:
+                if (!cap)
+                    return 0;
+                switch(propId)
```
Lets avoid multi-level switches.

If capture is not opened then any query calls are invalid (0 value is a legacy from C API).

tailsu replied:
Done, also applied to set().

tailsu commented on modules/videoio/src/cap_interface.hpp:
```diff
@@ -219,6 +219,31 @@ class IVideoCapture
     virtual int getCaptureDomain() { return CAP_ANY; } // Return the type of the capture object: CAP_DSHOW, etc...
+
+protected:
+    void rotateFrame(OutputArray mat) const
```
OK, done. Left it as an inline function, but I'd prefer to move the body to a .cpp file - just don't know which.

tailsu commented:
Added tests. I parametrized the ffmpeg tests by backend API, and moved them to test_orientation.cpp. I also changed an assert to check that both backends support orientation metadata.

asmorkalov commented:
The new test fails on Mac:
```
2022-11-21T12:50:40.9930830Z [ RUN      ] videoio/VideoCaptureAPITests.mp4_orientation_meta_auto/0, where GetParam() = FFMPEG
2022-11-21T12:50:41.0010960Z [       OK ] videoio/VideoCaptureAPITests.mp4_orientation_meta_auto/0 (14 ms)
2022-11-21T12:50:41.0011300Z [ RUN      ] videoio/VideoCaptureAPITests.mp4_orientation_meta_auto/1, where GetParam() = AVFOUNDATION
2022-11-21T12:50:41.0102350Z /Users/opencv-cn/GHA-OCV-2/_work/opencv/opencv/opencv/modules/videoio/test/test_orientation.cpp:41: Failure
2022-11-21T12:50:41.0102710Z Expected equality of these values:
2022-11-21T12:50:41.0102860Z   384
2022-11-21T12:50:41.0102980Z   frame.cols
2022-11-21T12:50:41.0103110Z     Which is: 0
2022-11-21T12:50:41.0103940Z [  FAILED  ] videoio/VideoCaptureAPITests.mp4_orientation_meta_auto/1, where GetParam() = AVFOUNDATION (9 ms)
```

asmorkalov commented:
CI platform:
```
2022-11-21T12:33:52.0293350Z OS: macOS Monterey 12.4
2022-11-21T12:33:52.0293600Z CPU:
2022-11-21T12:33:52.0295020Z 	 model: Intel(R) Core(R) i7 CPU @ 2.20GHz
2022-11-21T12:33:52.0295230Z 	 number of CPU: 1
2022-11-21T12:33:52.0295420Z 	 total cores: 6
2022-11-21T12:33:52.0295560Z GPU:
2022-11-21T12:33:52.0295790Z 	 model: Intel(R) UHD Graphics 630
2022-11-21T12:33:52.0295980Z 	 number of GPU: 1
2022-11-21T12:33:52.0296120Z RAM:
2022-11-21T12:33:52.0296270Z 	 size: 32GB
2022-11-21T12:33:52.0296720Z 	 type: DDR4-2667
```

asmorkalov commented:
I tried the rotated video and it works well with mplayer (ffmpeg) and vlc without hardware acceleration. VDPAU backend for VLC fails to play the video with rotation too. Not sure if it's related to rotation or very small amount of frames or some other option. Could you share very short clip with the option that could be used as test video?

tailsu commented:
Pushed a new video in https://github.com/opencv/opencv_extra/pull/1019

asmorkalov commented on modules/videoio/src/cap_interface.hpp:
```diff
@@ -221,6 +221,30 @@ class IVideoCapture
     virtual int getCaptureDomain() { return CAP_ANY; } // Return the type of the capture object: CAP_DSHOW, etc...
 };
 
+static void applyMetadataRotation(const IVideoCapture& cap, OutputArray mat)
```
`rotation_auto` and `rotation_angle` are enough as parameters. I propose to split concerns.

tailsu replied:
So, something like

```
 static void applyMetadataRotation(bool rotation_auto, int rotation_angle, OutputArray mat) {
  if (!rotation_auto) return;
  ...
}
```
and at the call site

```
applyMetadataRotation(
  0 != cap.getProperty(CAP_PROP_ORIENTATION_AUTO),
  static_cast<int>(cap.getProperty(CAP_PROP_ORIENTATION_META)), 
  image);
```
?

I don't what the call site looks like in this variation. Maybe split applyMetadataRotation in two:

```
static void applyMetadataRotation(const IVideoCapture& cap, OutputArray mat) {
  bool rotation_auto = 0 != cap.getProperty(CAP_PROP_ORIENTATION_AUTO);
  int rotation_angle = static_cast<int>(cap.getProperty(CAP_PROP_ORIENTATION_META));
  if (rotation_auto) {
    applyMetadataRotation(mat, rotation_angle);
  }
}

static void applyMetadataRotation(int rotation_angle, OutputArray mat) {
   ...
}
```

asmorkalov replied:
Yes, I mean `static void applyMetadataRotation(int rotation_angle, bool autorotation, OutputArray mat)`. The current version works for me too.

tailsu replied:
I'd prefer to keep it as-is, then. The next PR that adds support for orientation metadata to another backend can decide to refactor this function, I guess, if necessary.

asmorkalov reviewed (approved):
Looks good to me in general!

asmorkalov commented:
@alalek Could you take a look?

alalek commented on modules/videoio/src/cap_interface.hpp:
```diff
@@ -221,6 +221,30 @@ class IVideoCapture
     virtual int getCaptureDomain() { return CAP_ANY; } // Return the type of the capture object: CAP_DSHOW, etc...
 };
 
+static void applyMetadataRotation(const IVideoCapture& cap, OutputArray mat)
```
I would prefer to split definition (here) and declaration (e.g., in cap.cpp file)

tailsu replied:
Will do.

alalek reviewed (approved):
Looks good :+1:

alalek commented:
opencv_extra PR is not provided, not reviewed and not merged. Whole CI is RED.

mshabunin commented on modules/videoio/src/cap_ffmpeg.cpp:
```diff
@@ -112,7 +112,7 @@ class CvCapture_FFMPEG_proxy CV_FINAL : public cv::IVideoCapture
 
         cv::Mat tmp(height, width, CV_MAKETYPE(CV_8U, cn), data, step);
-        this->rotateFrame(tmp);
+        applyMetadataRotation(*this, tmp);
```
Apparently this PR breaks FFmpeg plugin build (with `-DVIDEOIO_PLUGIN_LIST=ffmpeg` option). In this mode it does not compile `cap.cpp` which contains `applyMetadataRotation` function and can not find it. Perhaps we could move this function to one of headers used by both AVFoundation and FFmpeg backends or to a separate source module and add it to the plugin compilation process.

tailsu replied:
Should I take care of this in a separate PR or will you do it internally in the team?

mshabunin replied:
I'll try to fix it today.
